File: src/shared/ai/neural/encode.ts
/**
 * Entity-based observation/action encoder (guide §3). Fixed-size padded
 * tensors with explicit masks; acting player is always relative index 0.
 * The Python training environment mirrors this exactly; the golden-vector
 * test (scripts/ai/neural/golden.ts -> ai_bot/neural/tests/golden_test.py)
 * proves parity.
 */

import { getCard, getNoble } from '../../data/gameData.js';
import type { AIObservation } from '../observation.js';
import type {
  DevelopmentCard,
  MainAction,
  TokenCounts,
} from '../../types/game.js';

export const MAX_PLAYERS = 4;
export const MAX_MARKET_CARDS = 12;
export const MAX_RESERVED_CARDS = 12;
export const MAX_NOBLES = 5;
export const MAX_ACTIONS = 64;

export const OBS_DIM = 462;
export const ACTION_DIM = 43;

const NORMAL = ['white', 'blue', 'green', 'red', 'black'] as const;
const ALL_TOKENS = [...NORMAL, 'gold'] as const;

const relSeat = (
  playerOrder: string[],
  selfID: string,
  targetID: string,
): number => {
  const selfIndex = playerOrder.indexOf(selfID);
  const targetIndex = playerOrder.indexOf(targetID);
  if (selfIndex < 0 || targetIndex < 0) return 0;
  return (targetIndex - selfIndex + playerOrder.length) % playerOrder.length;
};

const cardVector = (
  card: DevelopmentCard | undefined,
  present: number,
): number[] => [
  present,
  (card?.tier ?? 0) / 3,
  (card?.points ?? 0) / 5,
  ...NORMAL.map((color) => (card?.cost[color] ?? 0) / 7),
  ...NORMAL.map((color) => (card?.bonus === color ? 1 : 0)),
];

/** 462-dim observation vector. */
export const encodeObservation = (observation: AIObservation): Float32Array => {
  const out = new Float32Array(OBS_DIM);
  let cursor = 0;
  const push = (...values: number[]): void => {
    for (const value of values) out[cursor++] = value;
  };
  const selfID = observation.playerID;
  const playerOrder = observation.playerOrder;
  // Acting-player-relative order: self is always seat 0 (guide §3.1).
  const relativeOrder = [
    selfID,
    ...playerOrder.filter((playerID) => playerID !== selfID),
  ];

  // Global (15)
  push(...ALL_TOKENS.map((color) => observation.bank[color] / 7));
  push(
    observation.deckCounts[1] / 40,
    observation.deckCounts[2] / 40,
    observation.deckCounts[3] / 40,
  );
  push(
    observation.completedTurns / 200,
    observation.finalRound ? 1 : 0,
    observation.playerOrder.length / 4,
  );
  push(
    observation.pending === null ? 1 : 0,
    observation.pending?.type === 'discard' ? 1 : 0,
    observation.pending?.type === 'noble' ? 1 : 0,
  );

  // Players (4 x 16, fixed width regardless of presence)
  for (let seat = 0; seat < MAX_PLAYERS; seat += 1) {
    const playerID = relativeOrder[seat];
    const player = observation.players[playerID];
    if (!player) {
      push(...new Array(16).fill(0));
      continue;
    }
    push(...ALL_TOKENS.map((color) => player.tokens[color] / 10));
    const bonuses = NORMAL.map((color) =>
      player.purchasedCardIds.reduce(
        (sum, cardID) => sum + (getCard(cardID)?.bonus === color ? 1 : 0),
        0,
      ),
    );
    push(...bonuses.map((count) => count / 10));
    const score = player.purchasedCardIds.reduce(
      (sum, cardID) => sum + (getCard(cardID)?.points ?? 0),
      0,
    ) + player.nobleIds.reduce(
      (sum, nobleID) => sum + (getNoble(nobleID)?.points ?? 0),
      0,
    );
    push(
      score / 15,
      player.purchasedCardIds.length / 25,
      player.reservedCards.length / 3,
      player.nobleIds.length / 5,
      playerID === selfID ? 1 : 0,
    );
  }

  // Market cards (12 x 13)
  for (let tier = 1; tier <= 3; tier += 1) {
    for (const cardID of observation.market[tier as 1 | 2 | 3]) {
      push(...cardVector(cardID ? getCard(cardID) : undefined, cardID ? 1 : 0));
    }
  }

  // Reserved cards (12 x 16, fixed width regardless of presence)
  for (let seat = 0; seat < MAX_PLAYERS; seat += 1) {
    const playerID = relativeOrder[seat];
    const player = observation.players[playerID];
    for (let index = 0; index < 3; index += 1) {
      const reserved = player?.reservedCards[index];
      if (!reserved) {
        push(...new Array(16).fill(0));
        continue;
      }
      const known = reserved.cardId !== null;
      const card = known ? getCard(reserved.cardId ?? '') : undefined;
      push(
        1,
        relSeat(playerOrder, selfID, playerID) / MAX_PLAYERS,
        reserved.tier / 3,
        reserved.source === 'market' ? 1 : 0,
        known ? 1 : 0,
        (card?.points ?? 0) / 5,
        ...NORMAL.map((color) => (card?.cost[color] ?? 0) / 7),
        ...NORMAL.map((color) => (card?.bonus === color ? 1 : 0)),
      );
    }
  }

  // Nobles (5 x 7)
  for (let index = 0; index < MAX_NOBLES; index += 1) {
    const nobleID = observation.availableNobleIds[index];
    const noble = nobleID ? getNoble(nobleID) : undefined;
  push(
    nobleID ? 1 : 0,
    (noble?.points ?? 0) / 3,
    ...NORMAL.map((color) => (noble?.requirement[color] ?? 0) / 7),
  );
  }
  return out;
};

const actionTypeIndex = (
  action: MainAction | { type: 'discardTokens' | 'chooseNoble' },
): number => {
  switch (action.type) {
    case 'takeDifferent':
      return 0;
    case 'takeSame':
      return 1;
    case 'reserveMarket':
      return 2;
    case 'reserveDeck':
      return 3;
    case 'purchase':
      return 4;
    case 'pass':
      return 5;
    case 'discardTokens':
      return 6;
    case 'chooseNoble':
      return 7;
  }
};

const targetCardFor = (
  observation: AIObservation,
  action: MainAction,
): { card: DevelopmentCard | undefined; known: boolean } => {
  if (action.type === 'purchase') {
    if (action.location.source === 'market') {
      return { card: getCard(action.location.cardId), known: true };
    }
    const reserved = observation.players[observation.playerID]?.reservedCards.find(
      (entry) => entry.cardId === action.location.cardId,
    );
    return { card: getCard(action.location.cardId), known: true };
  }
  if (action.type === 'reserveMarket') {
    return { card: getCard(action.cardId), known: true };
  }
  return { card: undefined, known: false };
};

/** 43-dim action vector. */
export const encodeAction = (
  action: MainAction | { type: 'discardTokens'; tokens: TokenCounts } | {
    type: 'chooseNoble';
    nobleID: string;
  },
  observation: AIObservation,
): Float32Array => {
  const out = new Float32Array(ACTION_DIM);
  let cursor = 0;
  const push = (...values: number[]): void => {
    for (const value of values) out[cursor++] = value;
  };
  const typeIndex = actionTypeIndex(action);
  for (let index = 0; index < 8; index += 1) {
    push(typeIndex === index ? 1 : 0);
  }
  const phase =
    action.type === 'discardTokens'
      ? 1
      : action.type === 'chooseNoble'
        ? 2
        : 0;
  push(phase === 0 ? 1 : 0, phase === 1 ? 1 : 0, phase === 2 ? 1 : 0);
  let tier = 0;
  let tokenDelta = [0, 0, 0, 0, 0, 0];
  let payment = [0, 0, 0, 0, 0, 0];
  let target: { card: DevelopmentCard | undefined; known: boolean } = {
    card: undefined,
    known: false,
  };
  let noble = { points: 0, requirement: [0, 0, 0, 0, 0] };
  if (action.type === 'takeDifferent') {
    tokenDelta = NORMAL.map((color) => (action.colors.includes(color) ? 1 : 0));
  } else if (action.type === 'takeSame') {
    tokenDelta = NORMAL.map((color) => (color === action.color ? 2 : 0));
  } else if (action.type === 'reserveMarket') {
    tier = action.tier;
    target = targetCardFor(observation, action);
  } else if (action.type === 'reserveDeck') {
    tier = action.tier;
  } else if (action.type === 'purchase') {
    target = targetCardFor(observation, action);
    if (action.location.source === 'market') tier = action.location.tier;
    payment = ALL_TOKENS.map((color) => action.payment[color]);
  } else if (action.type === 'discardTokens') {
    payment = ALL_TOKENS.map((color) => action.tokens[color]);
  } else if (action.type === 'chooseNoble') {
    const nobleData = getNoble(action.nobleID);
    noble = {
      points: nobleData?.points ?? 0,
      requirement: NORMAL.map((color) => nobleData?.requirement[color] ?? 0),
    };
  }
  push(
    tier === 1 ? 1 : 0,
    tier === 2 ? 1 : 0,
    tier === 3 ? 1 : 0,
  );
  push(...tokenDelta.map((value) => value / 3));
  push(...payment.map((value) => value / 7));
  push(
    (target.card?.points ?? 0) / 5,
    ...NORMAL.map((color) => (target.card?.cost[color] ?? 0) / 7),
    ...NORMAL.map((color) => (target.card?.bonus === color ? 1 : 0)),
  );
  push(
    noble.points / 3,
    ...noble.requirement.map((value) => value / 7),
  );
  return out;
};

/** Convenience: encode a legal-action move from the AI candidate shape. */
export const encodeAIMove = (
  move: { move: 'mainAction' | 'discardTokens' | 'chooseNoble'; args: unknown[] },
  observation: AIObservation,
): Float32Array => {
  const [argument] = move.args;
  if (move.move === 'mainAction') {
    return encodeAction(argument as MainAction, observation);
  }
  if (move.move === 'discardTokens') {
    return encodeAction(
      { type: 'discardTokens', tokens: argument as TokenCounts },
      observation,
    );
  }
  return encodeAction(
    { type: 'chooseNoble', nobleID: argument as string },
    observation,
  );
};

File: ai_bot/neural/features/encode.py
"""Python mirror of src/shared/ai/neural/encode.ts (golden-vector parity)."""

from __future__ import annotations

from typing import Any, Dict, List, Optional

from env.game import get_card, get_noble, load_data

NORMAL = ["white", "blue", "green", "red", "black"]
ALL_TOKENS = NORMAL + ["gold"]

MAX_PLAYERS = 4
MAX_MARKET_CARDS = 12
MAX_RESERVED_CARDS = 12
MAX_NOBLES = 5
MAX_ACTIONS = 64

OBS_DIM = 462
ACTION_DIM = 43


def _rel_seat(player_order: List[str], self_id: str, target_id: str) -> int:
    try:
        self_index = player_order.index(self_id)
        target_index = player_order.index(target_id)
    except ValueError:
        return 0
    return (target_index - self_index + len(player_order)) % len(player_order)


def _card_vector(card: Optional[Dict[str, Any]], present: float) -> List[float]:
    return [
        present,
        (card["tier"] if card else 0) / 3.0,
        (card["points"] if card else 0) / 5.0,
    ] + [((card["cost"][color] if card else 0) / 7.0) for color in NORMAL] + [
        1.0 if card and card["bonus"] == color else 0.0 for color in NORMAL
    ]


def _bonuses_of(purchased_ids: List[str]) -> List[int]:
    counts = {color: 0 for color in NORMAL}
    for card_id in purchased_ids:
        card = get_card(card_id)
        if card:
            counts[card["bonus"]] += 1
    return [counts[color] for color in NORMAL]


def _score_of(purchased_ids: List[str], noble_ids: List[str]) -> int:
    card_points = sum(
        (get_card(card_id) or {}).get("points", 0) for card_id in purchased_ids
    )
    noble_points = sum(
        (get_noble(noble_id) or {}).get("points", 0) for noble_id in noble_ids
    )
    return card_points + noble_points


def _tier_section(observation: Dict[str, Any], key: str, tier: int) -> Any:
    """Observations from TS JSON use string tier keys; Python env uses ints."""
    section = observation[key]
    if tier in section:
        return section[tier]
    return section[str(tier)]


def encode_observation(observation: Dict[str, Any]) -> List[float]:
    out: List[float] = []
    self_id = observation["playerID"]
    player_order = observation["playerOrder"]
    relative_order = [self_id] + [
        player_id for player_id in player_order if player_id != self_id
    ]

    # Global (15)
    out.extend(observation["bank"][color] / 7.0 for color in ALL_TOKENS)
    out.extend(
        _tier_section(observation, "deckCounts", tier) / 40.0
        for tier in (1, 2, 3)
    )
    out.extend(
        [
            observation["completedTurns"] / 200.0,
            1.0 if observation["finalRound"] else 0.0,
            len(observation["playerOrder"]) / 4.0,
        ]
    )
    pending = observation["pending"]
    out.extend(
        [
            1.0 if pending is None else 0.0,
            1.0 if pending and pending["type"] == "discard" else 0.0,
            1.0 if pending and pending["type"] == "noble" else 0.0,
        ]
    )

    # Players (4 x 19)
    for seat in range(MAX_PLAYERS):
        player_id = relative_order[seat] if seat < len(relative_order) else None
        player = observation["players"].get(player_id) if player_id else None
        if not player:
            out.extend([0.0] * 16)
            continue
        out.extend(player["tokens"][color] / 10.0 for color in ALL_TOKENS)
        out.extend(count / 10.0 for count in _bonuses_of(player["purchasedCardIds"]))
        out.extend(
            [
                _score_of(player["purchasedCardIds"], player["nobleIds"]) / 15.0,
                len(player["purchasedCardIds"]) / 25.0,
                len(player["reservedCards"]) / 3.0,
                len(player["nobleIds"]) / 5.0,
                1.0 if player_id == self_id else 0.0,
            ]
        )

    # Market cards (12 x 13)
    for tier in (1, 2, 3):
        for card_id in _tier_section(observation, "market", tier):
            out.extend(
                _card_vector(get_card(card_id) if card_id else None, 1.0 if card_id else 0.0)
            )

    # Reserved cards (12 x 15)
    for seat in range(MAX_PLAYERS):
        player_id = relative_order[seat] if seat < len(relative_order) else None
        player = observation["players"].get(player_id) if player_id else None
        for index in range(3):
            reserved = player["reservedCards"][index] if player and index < len(player["reservedCards"]) else None
            if not reserved:
                out.extend([0.0] * 16)
                continue
            known = reserved["cardId"] is not None
            card = get_card(reserved["cardId"]) if known else None
            out.extend(
                [
                    1.0,
                    _rel_seat(player_order, self_id, player_id) / MAX_PLAYERS,
                    reserved["tier"] / 3.0,
                    1.0 if reserved["source"] == "market" else 0.0,
                    1.0 if known else 0.0,
                    (card["points"] if card else 0) / 5.0,
                ]
                + [((card["cost"][color] if card else 0) / 7.0) for color in NORMAL]
                + [1.0 if card and card["bonus"] == color else 0.0 for color in NORMAL]
            )

    # Nobles (5 x 7)
    for index in range(MAX_NOBLES):
        noble_id = (
            observation["availableNobleIds"][index]
            if index < len(observation["availableNobleIds"])
            else None
        )
        noble = get_noble(noble_id) if noble_id else None
        out.extend(
            [
                1.0 if noble_id else 0.0,
                (noble["points"] if noble else 0) / 3.0,
            ]
            + [((noble["requirement"][color] if noble else 0) / 7.0) for color in NORMAL]
        )
    assert len(out) == OBS_DIM, f"obs dim {len(out)} != {OBS_DIM}"
    return out


def _action_type_index(action_type: str) -> int:
    return {
        "takeDifferent": 0,
        "takeSame": 1,
        "reserveMarket": 2,
        "reserveDeck": 3,
        "purchase": 4,
        "pass": 5,
        "discardTokens": 6,
        "chooseNoble": 7,
    }[action_type]


def _target_card_for(
    observation: Dict[str, Any], action: Dict[str, Any]
) -> Optional[Dict[str, Any]]:
    if action["type"] == "purchase":
        return get_card(action["location"]["cardId"])
    if action["type"] == "reserveMarket":
        return get_card(action["cardId"])
    return None


def encode_action(
    action: Dict[str, Any], observation: Dict[str, Any]
) -> List[float]:
    out: List[float] = []
    type_index = _action_type_index(action["type"])
    out.extend(1.0 if index == type_index else 0.0 for index in range(8))
    phase = 1 if action["type"] == "discardTokens" else (2 if action["type"] == "chooseNoble" else 0)
    out.extend(
        [
            1.0 if phase == 0 else 0.0,
            1.0 if phase == 1 else 0.0,
            1.0 if phase == 2 else 0.0,
        ]
    )
    tier = 0
    token_delta = [0.0] * 6
    payment = [0.0] * 6
    target = None
    noble = {"points": 0, "requirement": [0] * 5}
    if action["type"] == "takeDifferent":
        token_delta = [
            1.0 if color in action["colors"] else 0.0 for color in NORMAL
        ] + [0.0]
    elif action["type"] == "takeSame":
        token_delta = [
            2.0 if color == action["color"] else 0.0 for color in NORMAL
        ] + [0.0]
    elif action["type"] == "reserveMarket":
        tier = action["tier"]
        target = _target_card_for(observation, action)
    elif action["type"] == "reserveDeck":
        tier = action["tier"]
    elif action["type"] == "purchase":
        target = _target_card_for(observation, action)
        if action["location"]["source"] == "market":
            tier = action["location"]["tier"]
        payment = [action["payment"][color] for color in ALL_TOKENS]
    elif action["type"] == "discardTokens":
        payment = [action["tokens"][color] for color in ALL_TOKENS]
    elif action["type"] == "chooseNoble":
        noble_data = get_noble(action["nobleID"])
        noble = {
            "points": noble_data["points"] if noble_data else 0,
            "requirement": [
                noble_data["requirement"][color] if noble_data else 0
                for color in NORMAL
            ],
        }
    out.extend(
        [
            1.0 if tier == 1 else 0.0,
            1.0 if tier == 2 else 0.0,
            1.0 if tier == 3 else 0.0,
        ]
    )
    out.extend(value / 3.0 for value in token_delta)
    out.extend(value / 7.0 for value in payment)
    out.extend(
        [
            (target["points"] if target else 0) / 5.0,
        ]
        + [((target["cost"][color] if target else 0) / 7.0) for color in NORMAL]
        + [1.0 if target and target["bonus"] == color else 0.0 for color in NORMAL]
    )
    out.extend(
        [noble["points"] / 3.0]
        + [value / 7.0 for value in noble["requirement"]]
    )
    assert len(out) == ACTION_DIM, f"action dim {len(out)} != {ACTION_DIM}"
    return out


def encode_ai_move(
    move: Dict[str, Any], observation: Dict[str, Any]
) -> List[float]:
    argument = move["args"][0]
    if move["move"] == "mainAction":
        return encode_action(argument, observation)
    if move["move"] == "discardTokens":
        return encode_action(
            {"type": "discardTokens", "tokens": argument}, observation
        )
    return encode_action({"type": "chooseNoble", "nobleID": argument}, observation)
